computron/computron.go

```diff
@@ -0,0 +1,438 @@
+	// extracted.
+	// [OPTIONAL] [DEFAULT: /tmp/synnax/computron]
+	PythonInstallDir string
+}
+
+var (
+	_ config.Config[Config] = Config{}
+	//DefaultConfig is the default configuration for the computron service.
+	DefaultConfig = Config{
+		PythonInstallDir: filepath.Join(os.TempDir(), "synnax", "computron"),
+	}
+)
+
+// Validate implements config.Config.
+func (c Config) Validate() error {
+	v := validate.New("computron")
+	validate.NotEmptyString(v, "PythonInstallDir", c.PythonInstallDir)
+	return v.Error()
+}
+
+// Override implements config.Config.
+func (c Config) Override(other Config) Config {
+	c.Instrumentation = override.Zero(c.Instrumentation, other.Instrumentation)
+	c.PythonInstallDir = override.String(c.PythonInstallDir, other.PythonInstallDir)
+	return c
+}
+
+type Interpreter struct {
+	cfg     Config
+	globals *C.PyObject
+}
+
+// lockThreadAndGIL locks the current thread and the Python Global Interpreter Lock (GIL).
+func lockThreadAndGIL() func() {
+	runtime.LockOSThread()
+	gilState := C.PyGILState_Ensure()
+	return func() {
+		C.PyGILState_Release(gilState)
+		runtime.UnlockOSThread()
+	}
+}
+
+// New creates a new embedded Python interpreter.
+func New(cfgs ...Config) (*Interpreter, error) {
+	cfg, err := config.New(DefaultConfig, cfgs...)
+	if err != nil {
+		return nil, err
+	}
+	cfg.L.Info("starting embedded Python service",
+		zap.String("install_dir", cfg.PythonInstallDir),
+		zap.String("version", targetPythonVersion),
+	)
+	s := &Interpreter{cfg: cfg}
+	if err = s.initPython(); err != nil {
+		return nil, err
+	}
+	err = s.initGlobals()
+	if err != nil {
+		return nil, err
+	}
+	cfg.L.Info("embedded Python service started successfully")
+	return s, nil
+}
+
+func (s *Interpreter) initPython() error {
+	// Check if the directory exists
+	installDir := filepath.Join(s.cfg.PythonInstallDir, "python_install")
+	if _, err := os.Stat(s.cfg.PythonInstallDir); err != nil && !os.IsNotExist(err) {
+		return errors.Wrapf(err, "failed to check Python installation directory")
+	}
+	// Read the contents of the VERSION file
+	contents, err := os.ReadFile(filepath.Join(installDir, "VERSION"))
+	v := string(contents)
+	v = strings.ReplaceAll(v, " ", "")
+	if err != nil && !os.IsNotExist(err) {
+		return errors.Wrapf(err, "failed to read Python version file")
+	}
+	if strings.Contains(v, targetPythonVersion) {
+		// Check if the version is the same as the embedded Python version. If so,
+		// everything is already set up and we can return early.
+		s.cfg.L.Debug("Python already installed. skipping installation")
+	} else {
+		s.cfg.L.Debug("extracting embedded Python installation. this may take a few seconds")
+		if err = xembed.Extract(
+			embeddedPython,
+			s.cfg.PythonInstallDir,
+			dirPerm,
+			filePerm,
+		); err != nil {
+			return errors.Newf("failed to extract embedded Python files: %v", err)
+		}
+		s.cfg.L.Debug("embedded Python installation extracted")
+	}
+
+	if runtime.GOOS == "windows" {
```

Comment:
Instead of evaluating this at runtime, use compiler directives with different string constants.